src/record/peer-record/index.js

```diff
@@ -58,8 +66,8 @@ class PeerRecord extends Record {
   /**
    * Returns true if `this` record equals the `other`.
    *
-   * @param {Record} other
-   * @returns {boolean}
+   * @param {PeerRecord} other
+   * @returns {other is Record}
```

Comment:
Since other is already required to be a `PeerRecord` instance which implements `Record` interface it is granted so this return type will have no effect. 

```suggestion
   * @returns {boolean}
```

Reply:
It errors with:

```
Property 'equals' in type 'PeerRecord' is not assignable to the same property in base type 'Record'.
  Type '(other: PeerRecord) => boolean' is not assignable to type '(other: any) => other is Record'.
    Signature '(other: PeerRecord): boolean' must be a type predicate.
```

Per https://github.com/libp2p/js-libp2p-interfaces/blob/feat/add-types/src/record/types.ts#L20

Should I just change the param to any?

Reply:
Ok so there are couple of things here:

1. That error occurs because interface says you can pass **anything** to the method, but then implementation contradicts it by saying by saying you only pass `PeerRecord`, implying that you can't pass `number` while interface says you can do that. In other words you can accept wider range of inputs (superset types) in implementations than type / interface requires and be compatible, but you can not do other way around.
2. I'm honestly confused by `equals(other: any): other is Record`. Using `other is Type` kind of annotations are only useful for predicates that are meant to be used for type refinements e.g. [example illustrating this](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgCIFcEGtkG8CwAUMsgI7qJYAUAlAFzIBuA9sACZEC+RRoksiFKmYBzPERIAjOFGr0mrDoW6EiCZiADOYZME0ZsyALzIqjOABt0EOnBABPeuaso9aTDiMA+CU0vXkADJA5DB7AAcIZhg-FwA6ckpjIxMAchh0EAQwYA1UnkIIAA9w5igddS0dAHcyrFARAFEiuABbcIsUEzN-G2ERAB8DLBpjL3FiXRiqPWGelxpRgkmSAHpV5ABJEFCACzdJC2ZDMMjkaNiAtwADYevfEmdrBIpsWl8VX2Bpp5Rg0IiURivxeSRSJgARBksjkNBClg9kOsSCjUWj0SQAHrY7HIAAKUGYkXK9mQqUS2FSyDYzAgmmQIGYOmKeh0GgBZ1Sw2QAzQolScWQiJBFLkiORKJxuIJROgYTJoqpNLpDKZyBZ2nOO1OKC5Hh5fJEAo+XCAA)
   ```ts
    interface Duck {
      quack(): void
    }

    interface Dog {
      bark(): void
    }

    const isDuck = (value:any):value is Duck =>
      value && typeof value.quack === 'function'

    export const workingExample = (value:Dog|Duck) => {
      if (isDuck(value)) {
        // In this block type of value is `Duck`
        value.quack()
      }

      if (value && typeof value.quack === "function") {
        //                      ^^^^^ Property 'quack' does not exist on type 'Duck | Dog'. 
        value.quack()
        //    ^^^^^ Property 'quack' does not exist on type 'Duck | Dog'.
      }
    }
   ```
   Note that in some cases on might restrict type of value in predicate in some way (does not need to be `any`), but the point is that TS essentially treats such predicate functions to do type refinements.

   I am not sure why one would want to do type refinement with `equals` check e.g. because if type of `value` in `value.equals(other)` is known and it's equal to `other` than `value` could be used in it's place e.g.:

   ```ts
   const example = (duck:Duck, other:Duck|Dog) {
     if (duck.equals(other)) {
       duck.quack() // as opposed to other.quack()
     }
   }
   ```

3. I think you may want to use `Record` for parameter type, as in it does not matter if `PeerRecord` implementation is used or some compatible `OtherRecord` implementation.

4. I think generally it is best to avoid `any` unless you want to tell TS to just shut up about that **value**. That is because in that case TS will not complain no matter you do with it. I think in most cases actual intent is to say that value is something of any type (via generics `interface Record { equals <T>(other:T):boolean }` or more simply that value is of `unknown` type (via `interface Record { equals(other:unknown):boolean }`). This inspired me to write [(what I hope is fun) explainer](https://www.typescriptlang.org/play?#code/MYGwhgzhAEBCD2APAPAFQHzQN4FgBQ00AbmCAK4CmAXNKvocPAHYQAuATmcK-OwBQly1WgEps9QtFYALAJYQAdIMrQAvMVKUJAXwlkmAIyR8RVVOIKT2FVmXZMpcxcoo78uvPllNWFdgDMwYApoAFEARwBBJgBPC0IAegToAHl-XwcABwp4TJAQsggQgAMwWOLoMqZ4VjBWWWZHCjj4dIoHAFsKMsc66AAqftlWQehgHoMQspiZbwBzCSToAHc5fKkY7LHpCmAAaz8pMAOYGTqAGmhMsHZpMEyIS554aBBh63AQOLKAEylrOowQbDUaQSqxWZMOYKCQUcJkUgQPjDGjTUzQIzwfJldz4PF4UCQGAANQAjGhMBREBkfjAECgMNBZB08hQuj4YBFonFcJYltE-qt2kyHDIQszWezavVGsCRv0macAawYGCIPAupC5tAINlgLJ-LJgNBkaKnGNIK4+ckwI5vCthtJoMUFPojIgTBVNdJ4D8xKxNiFgDt9ocILJ8qwvtB4JMYo9Km12JUQCAmawVmCeCAfotkjx-t0M3LQaqIXIoZUM6Q079ehnvKBKLTKnbK8tHc7XYZjCIvTYfT8YZY4QiQEiUWj4lYbHYzfJu+6TGpVOphove248B58EsAArsWP5DqK+un6oZqlBKNxVZ9MBkGS8aBdMqsS7VaB+Q-sVXWaDsGA8gUH8fSMCyEZWks9RdJc3hsN0fxinEEB2IEwRVgB+gwSEgC8G4ARHv4EwFDLNAZJ8AArCICijoilEiLuySSMxAB6bHsextCBqE7A-jQ65ukg541K2-j6NwDRMPi+LeL4ARBCEEQAMoagO8zTksADqdwZtkuTrIUY7Rq+LBSC8EBgHE8hniWCrjA4kytoM6revMoytOCcQBtksLwoi0DICprlQugyKsFQQVqVC6KYtiUnbvihJQGRABMFJftS7QtvS6USse7QqmE4SRVqGnJAAkg4qxGk64xFBsWzBn4BzJh2qZjBqeRAVVOwOLZp4cEWIGVEC-QuVFczuf4eY6nqBpGg1FCXG8Bw6qppVnBmdwwMUgken2L4Dr6w6ELR46BetFZzKFKIlVdYi8sx1i2PYdqKHty6rmurAbvtEiJExzFA8DIPQBxLHQAeuR+AG0AAOR7XD0A-PAFAwBeGXyBmjTeSEcN3fMcMKHwKUAMykwAnAxlgeDuBLgMlxKk+lVI0nSSC5Sy+UckVBOVo90BaXIwYOu12KgQ4D5PsmBYWTElyBP+7TwGQczSJc5XQHMIkxCryaZPAskKNA5X+AMQzysJ1bm+6gwzbrZAWkwcMZi6H19pcBiPisvXmyCCrWbaxT0hUYoOIwZA5tADtXNi9Vuz2+3FAA-AAhL5Y4QBdwXXWFNB83MD3-QBs6vcMIpsGUwQefS0AAGR129v2fau6bN9ThC09JeDEaRTP0TRfnjvRgvJIE45Wl4Ph+OhinhAAqkwezwMsDgC0slXQOhGZ3rpfjquHKtR3cRAhBZsh-B0vBn3lXk6YtMZm-18iXMsFCppcgz6Mvq+DMbGf+Tzt-aoq8YrwCxN0eKdMkokgACws0ykwbKHNGR5TZAVTkC8l4rzXjNTe1URZ1RCLjbYzVDhtTTOBLq9ohR9QtqMayg06jDSzFdaAewQGiheCfK0ANDpPj+LtROnoTpfiHhOcK0BgE4KLpYQgz05xN3diub67di5LFBpo1iHFIaHmyOwWGCNE5IxRmjaAGMqRYxjKKQM8NpGrzhluaBDMSQUQQWzOAKDMBoKlJgxeP9cHWmgNpGqos0zi3BJUR8PoZbmUspcB2cN-zsFkGrascxupMifvQgOo1v44NGCfeYM1vS+i7O7Co9lzEiUchwnBxtYDewNH7S2gcbZIHckwaMYdXg2CkTtduxQAHDxRFIpenDZFPVLg4cu8FahMGrmbWuDclHCLEF9Nu7snH4l7mRCifAADs1EzpIiOaPLeiIrS7OJPso5g9M58F2fSQ5IgxDQU4BQIAA)

If I had to guess you either want `equals(other:unknown): boolean` or bit more constrained `equals(other:Record):boolean` in both interface and a class.

Reply:
Thanks for the example and detailed go through. I am changing the interface + here to unknown and leverage the `instance of`